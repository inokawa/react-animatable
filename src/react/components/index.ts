export { TransitionGroup } from "./TransitionGroup";
export type { TransitionGroupProps } from "./TransitionGroup";
