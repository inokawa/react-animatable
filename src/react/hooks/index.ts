export * from "./useAnimation";
export * from "./useAnimationFunction";
export * from "./useTransitionAnimation";
export * from "./useScrollTimeline";
export * from "./useViewTimeline";
