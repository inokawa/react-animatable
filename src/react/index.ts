export * from "./components/index.js";
export * from "./hooks/index.js";
export * from "./types/index.js";
