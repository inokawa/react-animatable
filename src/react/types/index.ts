import {
  GetKeyframeFunction,
  TypedKeyframe,
  TypedKeyframeEffectOptions,
} from "../../core";
import { SCROLL_TIMELINE, VIEW_TIMELINE } from "./internal";

export interface AnimationOptions extends TypedKeyframeEffectOptions {
  timeline?: TimelineDefinition;
}

export type AnimationDefinition<Args> = [
  keyframe: TypedKeyframe | TypedKeyframe[] | GetKeyframeFunction<Args>,
  options?: AnimationOptions
];

export interface TimelineDefinition {
  (el: Element | null): void;
  /**
   * @private
   */
  _type: typeof SCROLL_TIMELINE | typeof VIEW_TIMELINE;
  /**
   * @private
   */
  _get: () => AnimationTimeline;
}

export type ScrollTimelineAxis = "block" | "inline" | "y" | "x";

export type ViewTimelineInset = "auto" | (string & {});
