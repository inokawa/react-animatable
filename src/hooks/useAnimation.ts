import { createRef, useEffect, useLayoutEffect, useRef, useState } from "react";
import { isSameObject, isSameObjectArray, toArray } from "../utils";
import {
  AnimationOptions,
  createAnimation,
  createHandle,
  TypedKeyframe,
} from "./core";

export type AnimationHandle = {
  play: () => AnimationHandle;
  replay: () => AnimationHandle;
  reverse: () => AnimationHandle;
  cancel: () => AnimationHandle;
  finish: () => AnimationHandle;
  pause: () => AnimationHandle;
  commit: () => AnimationHandle;
  setTime: (time: number) => AnimationHandle;
  setPlaybackRate: (
    rate: number | ((prevRate: number) => number)
  ) => AnimationHandle;
  end: () => Promise<void>;
};
export type WithRef<T> = T & { ref: React.RefObject<any> };

export const useAnimation = (
  keyframe: TypedKeyframe | TypedKeyframe[],
  options?: AnimationOptions
): WithRef<AnimationHandle> => {
  const keyframeRef = useRef(keyframe);
  const optionsRef = useRef(options);

  const [animation, cleanup] = useState<[WithRef<AnimationHandle>, () => void]>(
    () => {
      const ref = createRef<HTMLElement>();

      const getTarget = () => ref.current;
      const getKeyframes = () => toArray(keyframeRef.current);
      const getOptions = () => optionsRef.current;

      let cache:
        | [
            Animation,
            HTMLElement,
            TypedKeyframe[],
            AnimationOptions | undefined
          ]
        | undefined;
      const initAnimation = (
        keyframes: TypedKeyframe[],
        options: AnimationOptions | undefined
      ): Animation => {
        const el = getTarget()!;
        if (cache) {
          const [prevAnimation, prevEl, prevKeyframes, prevOptions] = cache;
          if (
            el === prevEl &&
            isSameObjectArray(keyframes, prevKeyframes) &&
            isSameObject(options, prevOptions)
          ) {
            return prevAnimation;
          }
          prevAnimation.cancel();
        }
        const animation = createAnimation(el, keyframes as Keyframe[], options);
        cache = [animation, el, keyframes, options];
        return animation;
      };
      const getAnimation = () => cache?.[0];
      const handle = createHandle();

      const externalHandle: WithRef<AnimationHandle> = {
        play: () => {
          handle._play(initAnimation(getKeyframes(), getOptions()));
          return externalHandle;
        },
        replay: () => {
          handle._replay(initAnimation(getKeyframes(), getOptions()));
          return externalHandle;
        },
        reverse: () => {
          handle._reverse(initAnimation(getKeyframes(), getOptions()));
          return externalHandle;
        },
        cancel: () => {
          handle._cancel(getAnimation());
          return externalHandle;
        },
        finish: () => {
          handle._finish(getAnimation());
          return externalHandle;
        },
        pause: () => {
          handle._pause(getAnimation());
          return externalHandle;
        },
        commit: () => {
          handle._commit(getAnimation(), getTarget()!, getKeyframes());
          return externalHandle;
        },
        setTime: (time) => {
          handle._setTime(getAnimation(), time);
          return externalHandle;
        },
        setPlaybackRate: (rate) => {
          handle._setRate(getAnimation(), rate);
          return externalHandle;
        },
        end: () => handle._end(getAnimation()),
        ref,
      };
      return [
        externalHandle,
        () => {
          handle._cancel(getAnimation());
        },
      ];
    }
  )[0];

  useLayoutEffect(() => {
    keyframeRef.current = keyframe;
    optionsRef.current = options;
  });

  useEffect(() => cleanup, []);

  return animation;
};
