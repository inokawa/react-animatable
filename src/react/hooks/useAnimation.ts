import { useEffect } from "react";
import { assign, getStyle, toArray } from "../../core/utils";
import {
  PlayOptions,
  _cancel,
  _waitFor,
  _finish,
  _pause,
  _play,
  _reverse,
  _setRate,
  _setTime,
  WaitingAnimationEventName,
  TypedKeyframe,
  GetKeyframeFunction,
} from "../../core/waapi";
import { useStatic } from "./useStatic";
import { useLatestRef } from "./useLatestRef";
import { AnimationDefinition } from "../types";
import {
  deleteAnimation,
  getAnimation,
  AnimationObject,
  initAnimation,
} from "./state";

const normalizeKeyframe = <Args>(
  el: Element,
  keyframe: TypedKeyframe | TypedKeyframe[] | GetKeyframeFunction<Args>,
  args: Args
): TypedKeyframe[] => {
  if (typeof keyframe === "function") {
    return keyframe(getStyle(el), args);
  }
  return toArray(keyframe);
};

export type PlayOptionsWithArgs<Args = void> = PlayOptions & { args: Args };

export interface BaseAnimationHandle<Args = void> {
  /**
   * A wrapper of Web Animations API's [play](https://developer.mozilla.org/en-US/docs/Web/API/Animation/play). It's similar to GSAP's `play()`.
   */
  play(
    ...opts: Args extends void ? [PlayOptions?] : [PlayOptionsWithArgs<Args>]
  ): BaseAnimationHandle<Args>;
  /**
   * A wrapper of Web Animations API's [reverse](https://developer.mozilla.org/en-US/docs/Web/API/Animation/reverse). It's similar to GSAP's `reverse()`.
   */
  reverse(): BaseAnimationHandle<Args>;
  /**
   * A wrapper of Web Animations API's [cancel](https://developer.mozilla.org/en-US/docs/Web/API/Animation/cancel). It's similar to GSAP's `kill()`.
   */
  cancel(): BaseAnimationHandle<Args>;
  /**
   * A wrapper of Web Animations API's [finish](https://developer.mozilla.org/en-US/docs/Web/API/Animation/finish).
   */
  finish(): BaseAnimationHandle<Args>;
  /**
   * A wrapper of Web Animations API's [pause](https://developer.mozilla.org/en-US/docs/Web/API/Animation/pause). It's similar to GSAP's `pause()`.
   */
  pause(): BaseAnimationHandle<Args>;
  /**
   * A setter of Web Animations API's [currentTime](https://developer.mozilla.org/en-US/docs/Web/API/Animation/currentTime). It's similar to GSAP's `seek()`.
   *
   * If you pass function, you can get [endTime](https://developer.mozilla.org/en-US/docs/Web/API/AnimationEffect/getComputedTiming#return_value) from its argument.
   */
  setTime(
    time: number | ((endTime: number) => number)
  ): BaseAnimationHandle<Args>;
  /**
   * A wrapper of Web Animations API's [updatePlaybackRate](https://developer.mozilla.org/en-US/docs/Web/API/Animation/updatePlaybackRate). It's similar to GSAP's `timeScale()`.
   *
   * If you pass function, you can get current [playbackRate](https://developer.mozilla.org/en-US/docs/Web/API/Animation/playbackRate) from its argument.
   */
  setPlaybackRate(
    rate: number | ((prevRate: number) => number)
  ): BaseAnimationHandle<Args>;
  /**
   * A getter of Promise that will be resolved in specified timing.
   *
   * - `finished`: resolved when animation is finished and its playback direction is normal.
   * - `reverseFinished`: resolved when animation is finished and its playback direction is reversed.
   */
  waitFor(event: WaitingAnimationEventName): Promise<BaseAnimationHandle<Args>>;
}

/**
 * Handle of {@link useAnimation}.
 * @typeParam Args - argument type
 */
export interface AnimationHandle<Args = void>
  extends BaseAnimationHandle<Args> {
  /**
   * You have to pass this callback to ref of element you want to control.
   */
  (ref: Element | null): void;
}

/**
 * A basic hook to use Web Animations API. See {@link AnimationHandle}.
 * @typeParam Args - argument type
 */
export const useAnimation = <Args = void>(
  ...args: AnimationDefinition<Args>
): AnimationHandle<Args> => {
  const argsRef = useLatestRef(args);

  const [handle, mount] = useStatic(
    (): [AnimationHandle<Args>, () => () => void] => {
      let _target: Element | null = null;
      let _active: AnimationObject | undefined;

      const init = (args: Args) => {
        if (!_target) return;
        const [keyframe, _options = {}] = argsRef.current;
        const { timeline, ...options } = _options;

        const prevActive = _active;

        return initAnimation(
          _target,
          (_active = {
            _keyframes: normalizeKeyframe(_target, keyframe, args),
            _options: options,
            _timeline: timeline?._get(),
          }),
          prevActive
        );
      };
      const clean = () => {
        if (_active) {
          deleteAnimation(_active);
        }
      };

      const externalHandle: AnimationHandle<Args> = assign(
        (ref: Element | null) => {
          if (!(_target = ref)) {
            clean();
          }
        },
        <BaseAnimationHandle<Args>>{
          play: (...opts) => {
            const animation = init(((opts[0] || {}) as { args?: Args }).args!);
            if (animation) {
              _play(animation);
            }
            return externalHandle;
          },
          reverse: () => {
            if (_active) {
              _reverse(getAnimation(_active));
            }
            return externalHandle;
          },
          cancel: () => {
            if (_active) {
              _cancel(getAnimation(_active));
            }
            return externalHandle;
          },
          finish: () => {
            if (_active) {
              _finish(getAnimation(_active));
            }
            return externalHandle;
          },
          pause: () => {
            if (_active) {
              _pause(getAnimation(_active));
            }
            return externalHandle;
          },
          setTime: (time) => {
            let animation = _active && getAnimation(_active);
            if (!animation) {
              const [keyframe] = argsRef.current;
              if (typeof keyframe === "function") {
                return externalHandle;
              }
              // Init animation in setTime to start animation without calling play
              animation = init(undefined!);
            }
            _setTime(animation, time);

            return externalHandle;
          },
          setPlaybackRate: (rate) => {
            if (_active) {
              _setRate(getAnimation(_active), rate);
            }
            return externalHandle;
          },
          waitFor: (event) => {
            return _waitFor(_active && getAnimation(_active), event).then(
              () => externalHandle
            );
          },
        }
      );

      return [externalHandle, () => clean];
    }
  );

  useEffect(mount, []);

  return handle;
};
