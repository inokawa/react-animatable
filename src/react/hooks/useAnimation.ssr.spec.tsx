/**
 * @vitest-environment node
 */
import { it, describe, expect } from "vitest";
import { useEffect } from "react";
import { renderToString, renderToStaticMarkup } from "react-dom/server";
import { useAnimation } from "./useAnimation";

describe("SSR", () => {
  const Comp = () => {
    const animate = useAnimation(
      { width: 100 },
      { duration: 600, easing: "ease-in-out" }
    );
    useEffect(() => {
      animate.play();
    }, []);
    return (
      <div ref={animate} style={{ width: 50 }}>
        test
      </div>
    );
  };

  it("should succeed with renderToString", () => {
    expect(renderToString(<Comp />)).toMatchSnapshot();
  });

  it("should succeed with renderToStaticMarkup", () => {
    expect(renderToStaticMarkup(<Comp />)).toMatchSnapshot();
  });
});
