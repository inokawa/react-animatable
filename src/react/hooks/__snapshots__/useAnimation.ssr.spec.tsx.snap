// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SSR should succeed with renderToStaticMarkup 1`] = `"<div style="width:50px">test</div>"`;

exports[`SSR should succeed with renderToString 1`] = `"<div style="width:50px">test</div>"`;
